docs/USAGE-GUIDE.md

```diff
@@ -1857,6 +1747,85 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
 [s2n_connection_wipe](#s2n\_connection\_wipe) does not need to be called prior to this function. **s2n_connection_free** performs its own wipe
 of sensitive data.
 
+## Private Key Operation Related Calls
+
+By default, s2n-tls automatically uses the configured private key to synchronously perform the signature
+and decryption operations required for a tls handshake. However, this default behavior may not
+work for some situations.
+
+For example:
+* An application may want to perform the CPU-expensive signature and decryption operations
+asynchronously to avoid blocking the main event loop.
+See [Asynchronous private key operations](#Asynchronous-private-key-operations)
+* An application may not have direct access to the private key, such as when using PKCS#11 or a TPM.
+See [Offloading private key operations](#Offloading-private-key-operations)
+
+To handle these use cases, s2n-tls provides a callback to allow users to control how these operations
+are performed. The callback is set via **s2n_config_set_async_pkey_callback** and is triggered 
+every time **s2n_negotiate** performs an action involving the private key. The callback is passed
+**op**, an opaque object representing the private key operation. To avoid memory leaks, **op** must
+always eventually be freed by calling **s2n_async_pkey_op_free**.
+
+The private key operation can be performed by calling **s2n_async_pkey_op_perform**
+(or **s2n_async_pkey_op_set_output**: see [Offloading private key operations](#Offloading-private-key-operations)).
+The required private key can be retrieved using the **s2n_connection_get_selected_cert** and **s2n_cert_chain_and_key_get_key** calls. The operation can then be finalized with **s2n_async_pkey_op_apply** to continue the handshake.
+
+### Asynchronous Private Key Operations
+
+When s2n-tls is used in non-blocking mode, private key operations can be completed
+asynchronously. This model can be useful to move execution of
+CPU-heavy private key operations out of the main
+event loop, preventing **s2n_negotiate** from blocking the loop for a few
+milliseconds each time the private key operation needs to be performed.
+
+To handle private key operations asynchronously, return from the callback without calling
+**s2n_async_pkey_op_perform** or **s2n_async_pkey_op_apply**. Usually the user would do this
+by spawning a separate thread to perform **op** and immediately returning **S2N_SUCCESS**
+from the callback without waiting for that separate thread to complete. In response,
+**s2n_negotiate** will return **S2N_FAILURE** with an error of type **S2N_ERR_T_BLOCKED**
+and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**.
+All subsequent calls to **s2n_negotiate** will produce the same result until **s2n_async_pkey_op_apply**
+is called to finalize the **op**.
```

Comment:
the distinction between `s2n_async_pkey_op_perform` and `s2n_async_pkey_op_apply` is not clear to me. is finalize here mean to clean up the memory or indicate that the operation has been applied?

why do we need both perform and apply?